Add Test coverage for XmlCommand.cs

SteveL-MSFT opened the pull request:
fixed issues in XmlCommand not working correctly for -First and -Skip
removed dead code form XmlCommand
added tests

I'll see if I can get some code coverage data

Fix https://github.com/PowerShell/PowerShell/issues/4152
<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to CONTRIBUTING.md.

-->

SteveL-MSFT committed:
fixed issues in XmlCommand not working correctly for -First and -Skip
removed dead code form XmlCommand
added tests

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -240,12 +240,6 @@ private void CreateFileStream()
-            // reset the read-only attribute
-            if (null != _readOnlyFileInfo)
-            {
-                _readOnlyFileInfo.Attributes |= FileAttributes.ReadOnly;
```
This is only setting this flag in memory and has no effect on the actual file

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -749,17 +737,10 @@ internal void Import()
                 while (!_deserializer.Done())
                 {
                     object result = _deserializer.Deserialize();
-                    if (item++ < skip)
```
this code is not necessary as ulong.MaxValue is a very large 64-bit number and this check is really only if it overflows

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -30,17 +30,17 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
 
-    It "Import with CliXml directive should work" -Skip:$IsOSX{
-		Get-Process | Export-Clixml $testfile
+ 	It "Import with CliXml directive should work" {
```
All the tests pass on Mac

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -49,4 +49,111 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
+		$one = [One]::New()
+		$one | Export-Clixml -Depth 2 -Path $testfile
+		$deserialized_one = Import-Clixml -Path $testfile
+		$deserialized_one.two.three.num | Should BeNullOrEmpty
```
Should also verify that ``` $deserialized_one.two.three ``` is not null

SteveL-MSFT replied:
Exception will be thrown if .three is null since I'm accessing a member

adityapatwardhan replied:
No exception is not thrown.

![image](https://user-images.githubusercontent.com/12820925/28030150-6170914c-6557-11e7-9692-c36ce3812f28.png)

SteveL-MSFT replied:
ok, will fix

JamesWTruher replied:
member access does not produce exception, unless strict mode is set. _Method_ invocation will throw null reference exception.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/xml.tests.ps1:
```diff
@@ -49,20 +49,55 @@
+        }
+
+        It "ToString() works correctly with file" {
+            $testfile = "$testdrive" + [System.IO.Path]::DirectorySeparatorChar + "test.xml"
```
Please use Join-Path

SteveL-MSFT replied:
will fix

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -49,4 +49,111 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
+		$ps.AddCommand("Get-Process")
+		$ps.AddCommand("Import-CliXml")
+		$ps.AddParameter("Path", $testfile)
+		$ps.BeginInvoke()
```
I can create a larger clixml file to import so that it's more likely StopProcessing() is called during operation and it hasn't completed.  Also noticed `Get-Process` shouldn't be there.

SteveL-MSFT replied:
Used @JamesWTruher code instead

SteveL-MSFT committed:
use join-path to create file paths
removed necessary get-process and increased content of xml to be imported to avoid race condition

fixed issues in XmlCommand not working correctly for -First and -Skip
removed dead code form XmlCommand

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -779,27 +760,30 @@ internal void Import()
 
-                    ICollection c = psObject.BaseObject as ICollection;
-                    if (c != null)
+                    if (psObject != null)
```
Minor note: The loop has unnecessary nesting and duplication.  Consider handling psObject != null and c != null) first to reduce nesting.  Something like the following:

object result = _deserializer.Deserialize();
PSObject psObject = result as PSObject;

if (psObject != null)
{
    ICollection c = psObject.BaseObject as ICollection;
    if (c != null)
    {
        foreach (object o in c)
        {
            if (count >= first)
                break;

            if (skipped++ >= skip)
            {
                count++;
                _cmdlet.WriteObject(o);
            }
        }
        continue;
    }
}

if (skipped++ >= skip)
{
    count++;
    _cmdlet.WriteObject(result);
}

SteveL-MSFT replied:
will fix

SteveL-MSFT commented:
@TravisEz13 the only thing I noticed was that my email wasn't public which I changed

SteveL-MSFT committed:
address PR feedback

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -49,4 +49,111 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
     }
+
+	It "Export-Clixml StopProcessing should succeed" {
+		$ps = [PowerShell]::Create()
```
there's some timing issues here, would the following suffice?
```PowerShell
$ps = [PowerShell]::Create()
$null = $ps.AddScript("1..10")
$null = $ps.AddCommand("foreach-object")
$null = $ps.AddParameter("Process", { $_; start-sleep 1 })
$null = $ps.AddCommand("Export-CliXml")
$null = $ps.AddParameter("Path", $testfile)
$null = $ps.BeginInvoke()
Start-Sleep 1
$null = $ps.Stop()
$ps.InvocationStateInfo.State | should be "Stopped"
$ps.Dispose()
```

SteveL-MSFT replied:
Thanks, I think that'll work better

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -49,4 +49,111 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
+		$out1 = Import-Clixml -Path $testfile	# this results in a hashtable
+		$out2 = Import-Clixml -Path $testfile -First 2 -Skip 1	# this results in a dictionary entry
+		$out2.Count | Should Be 2
+		$array = @()
```
an alternative approach to lines 126-135
```PowerShell
        ($out2.Name) -join ":" | should be (@($out1.Keys)[1,2] -join ":")
        ($out2.Value) -join ":" | should be (@($out1.Values)[1,2] -join ":")
```

SteveL-MSFT replied:
Got it, much simpler

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Xml.Tests.ps1:
```diff
@@ -49,5 +49,24 @@
+    }
+
+    # these tests just cover aspects that aren't normally exercised being used as a cmdlet
+	It "Can read back switch and parameter values using api" {
```
I don't think so.  The `ConvertTo-Xml` cmdlet has a getter for the `NoTypeInformation` parameter.  This is what I'm testing.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -30,17 +30,17 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
-    It "Import with CliXml directive should work" -Skip:$IsOSX{
-		Get-Process | Export-Clixml $testfile
+ 	It "Import with CliXml directive should work" {
+		Get-Process | Select-Object -First 5 | Export-Clixml $testfile
```
Good point.  I'll change this to `Get-Command` instead.

SteveL-MSFT committed:
simplified code and made more robust based on feedback

SteveL-MSFT commented:
@adityapatwardhan feedback addressed

SteveL-MSFT commented:
Finally successfully got updated code coverage numbers:

ConvertToXmlCommand 100%
ExportCliXmlCommand 100%
SelectXmlCommand 60.2% (the remainder are exception handling)
